corehq/apps/data_interfaces/models.py

```diff
@@ -1125,8 +1157,81 @@ def properties_fit_definition(self, case_properties):
         return all_match or any_match
 
     def when_case_matches(self, case, rule):
+        self._handle_case_duplicate_new(case, rule)
+        result = self._handle_case_duplicate(case, rule)
+
+        return result
+
+    def _handle_case_duplicate_new(self, case, rule, process_updates=False):
+        if not case_matching_rule_exists_in_es(case, rule):
+            # If the case isn't found, any duplicate information is unreliable
+            # Just queue this specific record up manually via resave_case
+            resave_case(rule.domain, case, send_post_save_signal=False)
+            return CaseRuleActionResult(num_updates=0)
+
+        try:
+            existing_duplicate = CaseDuplicateNew.objects.get(case_id=case.case_id, action=self)
+        except CaseDuplicateNew.DoesNotExist:
+            existing_duplicate = None
+
+        current_hash = CaseDuplicateNew.case_and_action_to_hash(case, self)
+
+        # Check if the new parameters have changed.
+        if existing_duplicate and existing_duplicate.match_values == current_hash:
+            # Nothing has changed. We can stop processing here
+            return CaseRuleActionResult(num_updates=0)
+
+        with transaction.atomic():
+            if existing_duplicate:
+                existing_duplicate.delete()
+            duplicate_ids = self._create_duplicates(case, rule, current_hash)
+
+        if process_updates:
+            num_updates = self._update_duplicates(duplicate_ids, case, rule)
+            return CaseRuleActionResult(num_updates=num_updates)
+        else:
+            return CaseRuleActionResult(num_updates=0)
+
+    def _create_duplicates(self, case, rule, current_hash):
+        # Handle whether or not this gets inserted as a new duplicate
+        matching_ids = find_matching_case_ids_in_es(case, rule, limit=3)
+
+        other_duplicate_ids = set([case_id for case_id in matching_ids if case_id != case.case_id])
+        if len(other_duplicate_ids) == 0:
+            # This isn't a duplicate, just return
+            return []
+
+        duplicates = [CaseDuplicateNew(case_id=case.case_id, action=self, match_values=current_hash)]
+        missing_ids = self._get_case_ids_not_recorded_as_duplicates(other_duplicate_ids)
+        if missing_ids:
+            # create a new duplicate for anything that currently isn't registered as one
+            new_duplicates = [
+                CaseDuplicateNew(case_id=missing_id, action=self, match_values=current_hash)
+                for missing_id in missing_ids
+            ]
+            duplicates.extend(new_duplicates)
```

Comment:
Can you clarify why this approach is preferred? Is the objection primarily that `if missing_ids:` is unnecessary? We could remove the `if` entirely and still `extend` the `duplicates` array (rather than create a for loop), but I liked the segmentation of the `if` statement

Reply:
The list comprehension allocates a new/temporary list, and then throws it away immediately after extending the `duplicates` list. It seems more direct/obvious to simply append to the `duplicates` list.

Reply:
Another option would be to skip the intermediate list
```py
        missing_ids = self._get_case_ids_not_recorded_as_duplicates(other_duplicate_ids)
        if missing_ids:
            # create a new duplicate for anything that currently isn't registered as one
            duplicates.extend(
                CaseDuplicateNew(case_id=missing_id, action=self, hash=current_hash)
                for missing_id in missing_ids
            )
```